### ISSUE TITLE
[BUG] Aztec code from D-Ticket not recognized

### ISSUE THREAD

**Issue by revuwa** [michaeltroger/greenpass-android#251]
**Describe the bug**
Thanks a lot for the experimental support of Aztec codes ️❤️ 
Am I the only one who have trouble getting the code of D-Tickets (Deutschland from Deutsche Bahn) recognized?
Even if I zoom in extremely, there is no chance to workaround.

**To Reproduce**
1. Open the app.
2. Add a D-Ticket (with Aztec code)
3. Zoom as much as you like, all day long
4. Scratch your head why the code could not be recognized

**Expected behavior**
Maybe it's just me or my understanding of what the app should do. But I thought, the (Aztec) code should be "extracted" and shown as a bigger version on top of app. But this not happens. It's just opens the D-Ticket and shows it as every other PDF app does.

**Screenshots**
![image](https://github.com/michaeltroger/greenpass-android/assets/7300276/75ba20fb-843c-4304-a485-4bff3a64ebcc)
_Of course the distortion was added later, for the screenshot._

**Smartphone (please complete the following information):**
 - Device: Pixel 8
 - OS: Android 14
 - App Version: 4.1.1

**Comment by michaeltroger** [comment]
Hi!
Thank you for reporting!
The ideal, expected behavior is that a document like:
https://github.com/michaeltroger/greenpass-android/blob/main/testdata/aztec.pdf
gets displayed like:
https://github.com/michaeltroger/greenpass-android/blob/main/screenshots/aztec_code.png

The zooming that you can do on the document has no impact on the detection though.

Personally I don't have any D-Ticket by hand. Therefore your screenshot was super valuable for trying to reproduce it.
I can reproduce no detection happening with an Aztec code holding 652 characters:
[aztec-dticket-demo_652.pdf](https://github.com/michaeltroger/greenpass-android/files/15226976/aztec-dticket-demo_652.pdf)

However the detection works if the code only holds 323 characters:
[aztec-dticket-demo_323.pdf](https://github.com/michaeltroger/greenpass-android/files/15226985/aztec-dticket-demo_323.pdf)

So I'm afraid the code on your ticket is either not sharp enough or holds too many signs to be easily detectable.
I have to say that this app doesn't re-invent the wheel and therefore uses the popular ZXing library for barcode detection. There are a few little tweaks that I can do, but not that much after all unfortunately. Aztec recognition isn't perfect in that library. The detection works on the Austrian train tickets at least and in the test document shared at the very top.

Since I don't have a D-Ticket by hand I'd be happy, if you could test something for me though:
Take a screenshot of the document and try if the code within the image gets recognized on https://zxing-cpp.github.io/zxing-cpp/demo_reader.html
if it doesn't get recognized, zoom in and create another screenshot -> try with that image. And so on.
If the code looks unsharp on the image then you could still try if another PDF reader app displays it sharper, and if it works when taking the screenshot from there.

In case detection works, then I would be interested in how much you had to zoom in and if you had any problems in terms of sharpness.

**Comment by michaeltroger** [comment]
With the attached fix the issue gets fixed for me.
Could you verify it for me?
Note: This setting (increased image resolution) would slow down the loading performance though.

If you're a programmer you can build from https://github.com/michaeltroger/greenpass-android/pull/252 this branch.
Alternatively you can install the test version from an APK file here: https://github.com/michaeltroger/greenpass-android/actions/runs/8976793372/artifacts/1478226889 (it's a zip file that first needs to get extracted)

**Comment by revuwa** [comment]
> With the attached fix the issue gets fixed for me. Could you verify that it works on the D-Ticket with this test version? Note: This setting (increased image resolution) would slow down the loading performance significantly though.
> 
> If you're a programmer then you can build the test app from [this branch](https://github.com/michaeltroger/greenpass-android/pull/252). Alternatively you can install the test version from an APK file here: https://github.com/michaeltroger/greenpass-android/actions/runs/8976793372/artifacts/1478226889 (it's a zip file that first needs to get extracted)
> 
> Please note that installing a test version (debug build) of the app, will make you loose all your app's data.

WOW, thank you so much for the super fast response and your enthusiasm 💕

Of course I would have been happy to help with troubleshooting/testing, but it seems that this is no longer necessary, because the debug build (mentioned above) solves the problem for me. 
So I can confirm that the [PR #252](https://github.com/michaeltroger/greenpass-android/pull/252) works for my current D-Ticket and just have to say thank you so much for all of your efforts (and your great app, by the way)!

**Comment by revuwa** [comment]
PS: the enlarged barcode (within your debug build) could be read without any problems during a passenger check, today.

If increasing the value of PDF_RESOLUTION_MULTIPLIER has the downside of slower performance, what do you think about an option for this? As long as it doesn't make things too complex, unnecessarily?

I would think about something like this:
![image](https://github.com/michaeltroger/greenpass-android/assets/7300276/a3462d56-0922-4beb-94e7-bcab6c2dc30e)
(This is just an edited image for illustrative purposes, I don't have coded anythin').

**Comment by michaeltroger** [comment]
Thanks for trying it out right away! Glad to hear that the fix works.

Yes, I thought about the same. A setting is definitely an option. 
I would not want to ship the change as it is just yet though. I still want to check if there is some optimization possible. Also there is a bit a weird loading state when fastly scrolling through PDFs with many pages. That bug is not directly related to the one here, but it becomes more noticeable with the longer loading time.

I'm busy these days but I'm definitely planning to ship this fix in the upcoming days/weeks in some way!

**Comment by michaeltroger** [comment]
The fix will be part of the next release.

I found a way to optimize the performance.
Basically I'm using the high resolution image only for barcode search now. For display a smaller resolution is used.
On top of that I added a memory cache for the barcodes/pdf pages.

With this the UI feels fluent now, but together with some other barcode detection improvements done earlier (for DataMatrix detection), it takes 6 times longer to compute compared to much earlier app versions. That's why I added a setting as well, it will be enabled by default though.
